Background WebContent Processes should fully suspend behind runtime flag

salinas-miguel opened the pull request:
#### 4bd642c8073d7a9880db59403458169ccd72183e
<pre>
Background WebContent Processes should fully suspend behind runtime flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=246304">https://bugs.webkit.org/show_bug.cgi?id=246304</a>
rdar://101001791

Reviewed by Chris Dumez.

Add a runtime flag &quot;FullySuspendBackgroundTabs&quot; which allows
ProcessThrottler to drop all RunningBoard assertions and consequently
suspend the throttled background WebContent process.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setAssertionTypeForTesting:]):
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::addActivity):
(WebKit::ProcessThrottler::removeActivity):
(WebKit::ProcessThrottler::expectedThrottleState):
(WebKit::ProcessThrottler::updateThrottleStateNow):
(WebKit::ProcessThrottler::assertionTypeForState):
(WebKit::ProcessThrottler::setThrottleState):
(WebKit::ProcessThrottler::updateThrottleStateIfNeeded):
(WebKit::ProcessThrottler::didConnectToProcess):
(WebKit::ProcessThrottler::prepareToSuspendTimeoutTimerFired):
(WebKit::ProcessThrottler::processReadyToSuspend):
(WebKit::ProcessThrottler::sendPrepareToSuspendIPC):
(WebKit::ProcessThrottler::enableFullSuspension):
(WebKit::ProcessThrottler::expectedAssertionType): Deleted.
(WebKit::ProcessThrottler::updateAssertionTypeNow): Deleted.
(WebKit::ProcessThrottler::setAssertionType): Deleted.
(WebKit::ProcessThrottler::updateAssertionIfNeeded): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/ProcessThrottlerClient.h:
(WebKit::ProcessThrottlerClient::didChangeThrottleState):
(WebKit::ProcessThrottlerClient::didSetAssertionType): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createWebPage):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didChangeThrottleState):
(WebKit::WebProcessProxy::didSetAssertionType): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::setAssertionTypeForTesting):

Canonical link: <a href="https://commits.webkit.org/255498@main">https://commits.webkit.org/255498@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58bbc7901b454831e741bf79d0977cd899cdc93f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102324 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96584 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1800 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30164 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84988 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98487 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98246 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1212 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79081 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28143 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83130 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82832 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71233 "Found 2 new API test failures: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive, /TestWTF:WTF_WorkQueue.DestroyDispatchedOnDispatchQueue (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83955 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36574 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16760 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78994 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34359 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17939 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27386 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3811 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38229 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40552 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81614 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40138 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37098 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18458 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -289,6 +312,11 @@ void ProcessThrottler::setAllowsActivities(bool allow)
 
+void ProcessThrottler::enableFullSuspension()
+{
+    m_fullySuspend = true;
```
Per WebKit coding style, boolean variables need a prefix. Here probably `m_shouldFullySuspend`.

salinas-miguel replied:
Thanks, fixed

cdumez replied:
But I am still confused by the "fully suspend" naming :)

salinas-miguel replied:
🤦‍♂️thanks fixed now

cdumez commented on Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
```diff
@@ -353,6 +353,18 @@ FullScreenEnabled:
     WebCore:
       default: false
 
+FullySuspendBackgroundTabs:
```
I find this naming confusing. Maybe it should be about taking Suspended assertions or not?
E.g. "ShouldTakeSuspendedAssertions" and make it true by default?

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.h:
```diff
@@ -52,6 +52,11 @@ typedef RefCounter<ProcessSuppressionDisabledCounterType> ProcessSuppressionDisa
 typedef ProcessSuppressionDisabledCounter::Token ProcessSuppressionDisabledToken;
 
 enum class IsSuspensionImminent : bool { No, Yes };
+enum class ProcessThrottleState {
```
`enum class ProcessThrottleState : uint8_t {`

salinas-miguel committed:
Background WebContent Processes should fully suspend behind runtime flag
https://bugs.webkit.org/show_bug.cgi?id=246304
rdar://101001791

Reviewed by Chris Dumez.

Add a runtime flag "FullySuspendBackgroundTabs" which allows
ProcessThrottler to drop all RunningBoard assertions and consequently
suspend the throttled background WebContent process.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setAssertionTypeForTesting:]):
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::addActivity):
(WebKit::ProcessThrottler::removeActivity):
(WebKit::ProcessThrottler::expectedThrottleState):
(WebKit::ProcessThrottler::updateThrottleStateNow):
(WebKit::ProcessThrottler::assertionTypeForState):
(WebKit::ProcessThrottler::setThrottleState):
(WebKit::ProcessThrottler::updateThrottleStateIfNeeded):
(WebKit::ProcessThrottler::didConnectToProcess):
(WebKit::ProcessThrottler::prepareToSuspendTimeoutTimerFired):
(WebKit::ProcessThrottler::processReadyToSuspend):
(WebKit::ProcessThrottler::sendPrepareToSuspendIPC):
(WebKit::ProcessThrottler::enableFullSuspension):
(WebKit::ProcessThrottler::expectedAssertionType): Deleted.
(WebKit::ProcessThrottler::updateAssertionTypeNow): Deleted.
(WebKit::ProcessThrottler::setAssertionType): Deleted.
(WebKit::ProcessThrottler::updateAssertionIfNeeded): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/ProcessThrottlerClient.h:
(WebKit::ProcessThrottlerClient::didChangeThrottleState):
(WebKit::ProcessThrottlerClient::didSetAssertionType): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createWebPage):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didChangeThrottleState):
(WebKit::WebProcessProxy::didSetAssertionType): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::setAssertionTypeForTesting):

Canonical link: https://commits.webkit.org/255498@main
